pkg/deviceplugins/device_manager.go

```diff
@@ -0,0 +1,502 @@
+)
+
+const (
+	vfioDevicePath    = "/dev/vfio/"
+	vfioMount         = "/dev/vfio/vfio"
+	pciBasePath       = "/sys/bus/pci/devices"
+	connectionTimeout = 120 * time.Second // Google gRPC default timeout
+	PCIResourcePrefix = "PCI_RESOURCE"
+)
+
+type PCIDevice struct {
+	pciID      string
+	driver     string
+	pciAddress string
+	iommuGroup string
+	numaNode   int
+}
+
+type PCIDevicePlugin struct {
+	pcidevs       []*PCIDevice
+	devs          []*pluginapi.Device
+	server        *grpc.Server
+	socketPath    string
+	stop          <-chan struct{}
+	health        chan deviceHealth
+	devicePath    string
+	resourceName  string
+	done          chan struct{}
+	deviceRoot    string
+	iommuToPCIMap map[string]string
+	initialized   bool
+	lock          *sync.Mutex
+	deregistered  chan struct{}
+	starter       *DeviceStarter
+}
+
+type DeviceStarter struct {
+	started  bool
+	stopChan chan struct{}
+	backoff  []time.Duration
+}
+
+func (dp *PCIDevicePlugin) GetPCIDevices() []*PCIDevice {
+	return dp.pcidevs
+}
+
+func (d *PCIDevice) GetID() string {
+	return d.pciID
+}
+
+func NewPCIDevicePlugin(pciDevices []*PCIDevice, resourceName string) *PCIDevicePlugin {
+	serverSock := SocketPath(strings.Replace(resourceName, "/", "-", -1))
+	iommuToPCIMap := make(map[string]string)
+
+	initHandler()
+
+	devs := constructDPIdevices(pciDevices, iommuToPCIMap)
+	dpi := &PCIDevicePlugin{
+		pcidevs:       pciDevices,
+		devs:          devs,
+		socketPath:    serverSock,
+		resourceName:  resourceName,
+		devicePath:    vfioDevicePath,
+		deviceRoot:    util.HostRootMount,
+		iommuToPCIMap: iommuToPCIMap,
+		health:        make(chan deviceHealth),
+		initialized:   false,
+		lock:          &sync.Mutex{},
+		starter: &DeviceStarter{
+			started:  false,
+			stopChan: make(chan struct{}),
+			backoff:  defaultBackoffTime,
+		},
+	}
+	return dpi
+}
+
+func constructDPIdevices(pciDevices []*PCIDevice, iommuToPCIMap map[string]string) (devs []*pluginapi.Device) {
+	for _, pciDevice := range pciDevices {
+		iommuToPCIMap[pciDevice.iommuGroup] = pciDevice.pciAddress
+		dpiDev := &pluginapi.Device{
+			ID:     string(pciDevice.pciID),
+			Health: pluginapi.Unhealthy,
+		}
+		if pciDevice.numaNode >= 0 {
+			numaInfo := &pluginapi.NUMANode{
+				ID: int64(pciDevice.numaNode),
+			}
+			dpiDev.Topology = &pluginapi.TopologyInfo{
+				Nodes: []*pluginapi.NUMANode{numaInfo},
+			}
+		}
+		devs = append(devs, dpiDev)
+	}
+	return
+}
+
+var defaultBackoffTime = []time.Duration{1 * time.Second, 2 * time.Second, 5 * time.Second, 10 * time.Second}
+
+func (dpi *PCIDevicePlugin) StartWithRetryAndBackOff() (err error) {
+	c := dpi.starter
+	if c.started {
+		return nil
+	}
+
+	stop := make(chan struct{})
+
+	logger := log.DefaultLogger()
+	dev := dpi
+	deviceName := dev.GetDeviceName()
+	logger.Infof("Starting a device plugin for device: %s", deviceName)
+	retries := 0
+
+	backoff := c.backoff
+
+	go func() {
```

Comment:
we could use [errgrp](https://pkg.go.dev/golang.org/x/sync/errgroup), and return the error from the go routine. This could be handled by the pcideviceclaim controller, and we could just use the k8s exponential back off rather than writing our own logic